### PR TITLE
Add a pool implementation

### PR DESCRIPTION
**Pull request by vinistock**
This PR adds a pool implementation to keep data avoiding duplication, the need of cloning (other than reference cloning) and type-safe IDs.

We can use a data structure like this one for URIs, names or whatever else we need to store and it will not allow for duplicate entries to be inserted. Other data structures can store usize IDs exclusively and not have to keep cloned strings around.

Additionally, by setting an empty type as the ID used for each pool, we can avoid simple mistakes during development like accidentally using an ID from one pool to index another one.

### CODE REVIEW

File: src/pool.rs [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+/// # let mut name_pool = Pool::<NameId, String>::new();
+/// # let mut uri_pool = Pool::<UriId, String>::new();
+/// # let name_id = name_pool.add("Foo::Bar#method".into());
+/// # uri_pool.get(&name_id) // Not possible. This will fail to compile because of the ID types
+/// ```
+pub struct Pool<I, T: Eq + std::hash::Hash> {
+    id_to_entry: HashMap<usize, Rc<T>>,
+    entry_to_id: HashMap<Rc<T>, usize>,
+    next_id: usize,
+    _phantom: PhantomData<I>,
```
Comment (ChallaHalla) [inline src/pool.rs:29]: What's the purpose of the `PhantomData` here?

File: src/pool.rs [hunk 2]
```diff
@@ -0,0 +1,132 @@
...
+/// # pub struct UriId;
+/// #
+/// # let mut name_pool = Pool::<NameId, String>::new();
+/// # let mut uri_pool = Pool::<UriId, String>::new();
+/// # let name_id = name_pool.add("Foo::Bar#method".into());
+/// # uri_pool.get(&name_id) // Not possible. This will fail to compile because of the ID types
+/// ```
+pub struct Pool<I, T: Eq + std::hash::Hash> {
+    id_to_entry: HashMap<usize, Rc<T>>,
+    entry_to_id: HashMap<Rc<T>, usize>,
```
Comment (ChallaHalla) [inline src/pool.rs]: I haven't taken a look at the parallel indexing PR yet. But for parallelizing access to the Pool, would we need to replace the `Rc` with `Arc` so that they are atomic ref counters?

File: src/pool.rs [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+        let mut pool: Pool<UniqueNameId, String> = Pool::new();
+        let id = pool.add("Foo::Bar#method".into());
+        let item = pool.get(&id).unwrap();
+        assert_eq!(*item, "Foo::Bar#method".to_string());
+    }
+
+    #[test]
+    fn test_adding_the_same_entry_just_returns_id() {
+        let mut pool: Pool<UniqueNameId, String> = Pool::new();
+        let id = pool.add("Foo::Bar#method".into());
```
Comment (ChallaHalla) [inline src/pool.rs:130]: Adding this more to solidify my own Rust knowledge so feel free to ignore.

Is the `.into()` here essentially cloning the String literal on the stack to the heap? Would it be more explicit to call `.clone()` in such situations?

File: src/pool.rs [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+/// ```
+pub struct Pool<I, T: Eq + std::hash::Hash> {
+    id_to_entry: HashMap<usize, Rc<T>>,
+    entry_to_id: HashMap<Rc<T>, usize>,
+    next_id: usize,
+    _phantom: PhantomData<I>,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct PoolId<I: PartialEq> {
```
Comment (Morriar) [inline src/pool.rs:33]: Do we really need this bound?

```suggestion
pub struct PoolId<I> {
```

File: src/pool.rs [hunk 5]
```diff
@@ -0,0 +1,132 @@
...
+/// # uri_pool.get(&name_id) // Not possible. This will fail to compile because of the ID types
+/// ```
+pub struct Pool<I, T: Eq + std::hash::Hash> {
+    id_to_entry: HashMap<usize, Rc<T>>,
+    entry_to_id: HashMap<Rc<T>, usize>,
+    next_id: usize,
+    _phantom: PhantomData<I>,
+}
+
+#[derive(PartialEq, Debug)]
```
Comment (Morriar) [inline src/pool.rs:32]: I wonder if we should add a few more derives here:

```suggestion
#[derive(Copy, Clone, PartialEq, Eq, Debug, Hash)]
```

File: src/pool.rs [hunk 6]
```diff
@@ -0,0 +1,132 @@
...
+        self.entry_to_id.insert(item_rc, id);
+        self.next_id += 1;
+        id.into()
+    }
+
+    pub fn get(&self, id: &PoolId<I>) -> Option<&T> {
+        self.id_to_entry.get(&id.value).map(|rc| rc.as_ref())
+    }
+
+    pub fn remove(&mut self, id: &PoolId<I>) {
```
Comment (Morriar) [inline src/pool.rs]: Should we make the `remove` return the value or `None`?

```suggestion
    pub fn remove(&mut self, id: &PoolId<I>) -> option<T> {
```

File: src/pool.rs [hunk 7]
```diff
@@ -0,0 +1,132 @@
...
+/// # #[derive(PartialEq, Debug)]
+/// # pub struct UriId;
+/// #
+/// # let mut name_pool = Pool::<NameId, String>::new();
+/// # let mut uri_pool = Pool::<UriId, String>::new();
+/// # let name_id = name_pool.add("Foo::Bar#method".into());
+/// # uri_pool.get(&name_id) // Not possible. This will fail to compile because of the ID types
+/// ```
+pub struct Pool<I, T: Eq + std::hash::Hash> {
+    id_to_entry: HashMap<usize, Rc<T>>,
```
Comment (Morriar) [inline src/pool.rs]: Do we expect more than 4 billions entries in the same pool? Could we use a `u32` instead?

```suggestion
    id_to_entry: HashMap<u32, Rc<T>>,
```

File: src/pool.rs [hunk 8]
```diff
@@ -0,0 +1,132 @@
...
+    pub struct OtherId;
+
+    #[test]
+    fn test_adding_and_fetching_entries() {
+        let mut pool: Pool<UniqueNameId, String> = Pool::new();
+        let id = pool.add("Foo::Bar#method".into());
+        let item = pool.get(&id).unwrap();
+        assert_eq!(*item, "Foo::Bar#method".to_string());
+    }
+
```
Comment (Morriar) [inline src/pool.rs:119]: Let's add a test showing retrieving an invalid ID returns `None`?

File: src/pool.rs (hunk 2 above)
Comment (Morriar) [reply src/pool.rs]: I agree, `Arc` may be better.

Or I wonder if we could make this generic: 

```rust
pub struct Pool<I, T, S>
where
    T: Eq + Hash,
    S: Clone + Deref<Target = T> + Eq + Hash,
{
    id_to_entry: HashMap<usize, S>,
    entry_to_id: HashMap<S, usize>,
    next_id: usize,
    _phantom: PhantomData<I>,
}
```

So the client can choose:

```rust
Pool<NameId, String, Rc<String>>
```

And provide type aliases:

```rust
pub type RcPool<I, T> = Pool<I, T, Rc<T>>;
pub type ArcPool<I, T> = Pool<I, T, Arc<T>>;
```

File: src/pool.rs (hunk 1 above)
Comment (Morriar) [inline src/pool.rs:29]: I wonder if `_marker` would be more idiomatic?

Comment (Morriar) [reply src/pool.rs:29]: Good read: https://doc.rust-lang.org/nomicon/phantom-data.html

Comment (vinistock) [reply src/pool.rs:29]: ☝️ highly recommend the read, but the idea is that if you simply add a type parameter `T` to a struct without using it, Rust is not happy. The `PhantomData` thing allows you to "use" the parameter `T` without actually using it.

It's basically a trick for the type checker to allow us differentiating based on `T`, but without actually having to add more data to the struct.

File: src/pool.rs (hunk 3 above)
Comment (vinistock) [reply src/pool.rs:130]: That's a good question and I'm not completely sure. I assumed that this would be equivalent to `to_string()`, since we don't accept string literals, but only strings in the heap.

File: src/pool.rs (hunk 2 above)
Comment (vinistock) [reply src/pool.rs]: I switched to using `Arc` for now. It's a bit early to tell if we're going to need an `Rc` and an `Arc` version of the pool, so I'd vote for us to revisit later if we realize there's a need.

File: src/pool.rs (hunk 4 above)
Comment (vinistock) [reply src/pool.rs:33]: It's needed to compare the IDs, otherwise Rust doesn't know that they can be checked with `==`.

File: src/pool.rs (hunk 5 above)
Comment (vinistock) [reply src/pool.rs:32]: I'd vote for us to add more as we realize we need them.

File: src/pool.rs (hunk 7 above)
Comment (vinistock) [reply src/pool.rs]: Not really, I was just trying to avoid conversions because vectors use `usize` for indices. I changed to u32.

File: src/pool.rs (hunk 1 above)
Comment (vinistock) [reply src/pool.rs:29]: I don't know to be honest, but I switched to `_marker`.

File: src/pool.rs (hunk 6 above)
Comment (vinistock) [reply src/pool.rs]: I don't know if we'll need it, but I added it as it better matches the behaviour of hashmaps and vectors.

File: src/pool.rs [hunk 9]
```diff
@@ -0,0 +1,145 @@
...
+/// ```
+pub struct Pool<I, T: Eq + Debug + Hash> {
+    id_to_entry: HashMap<u32, Arc<T>>,
+    entry_to_id: HashMap<Arc<T>, u32>,
+    next_id: u32,
+    _phantom: PhantomData<I>,
+}
+
+#[derive(PartialEq, Debug)]
+pub struct PoolId<I: PartialEq> {
```
Comment (ChallaHalla) [inline src/pool.rs:33]: Is there a reason we wrap the u32 ID value in a struct VS just copying around the u32 ID as needed? My guess is that there's some advantage to passing around the reference instead of copying the value constantly -- is that correct?

Comment (vinistock) [reply src/pool.rs:33]: The only goal is to leverage the type system to avoid accidentally trying to access a pool using an index that belongs to a different pool.

For example:

```rust
// Say we need to remember unique entry names and unique URIs we have indexed
let mut name_pool = Pool::<NameId, String>::new();
let mut uri_pool = Pool::<UriId, String>::new();

// Entries and other entities will store IDs and not the strings that they refer to
let name_id = name_pool.add("Foo::Bar#method".into());

// The empty structs that wrap the u32 values provide us type safety that we cannot
// accidentally use an ID from one pool to index into another pool. This will fail
// type checking and prevent compilation:
uri_pool.get(&name_id)
```

File: src/pool.rs (hunk 4 above)
Comment (Morriar) [reply src/pool.rs:33]: We need the _derive_ on `PartialEq` but not the _bound_ right?
